File: packages/dvai-bridge-capacitor-llama/ios/Tests/DVAICapacitorLlamaTests/ContentPartsTranslatorTest.swift
import XCTest
@testable import DVAICapacitorLlama

final class ContentPartsTranslatorTest: XCTestCase {
    // MARK: - Mocks

    /// Image decoder that returns canned bytes per URL. Records every call so
    /// tests can assert which URLs were passed in and in what order.
    final class MockImageDecoder: ImageDecoderProtocol {
        var responses: [String: Data] = [:]
        var calls: [String] = []
        func resolve(url: String) async throws -> Data {
            calls.append(url)
            if let bytes = responses[url] { return bytes }
            return Data([0xDE, 0xAD, 0xBE, 0xEF])
        }
    }

    /// Audio-decoder closure factory. Records each call's `(bytesIn, format)`
    /// and returns canned PCM bytes.
    final class AudioRecorder {
        var calls: [(Data, AudioFormat)] = []
        var pcmOut: Data = Data([0x11, 0x22, 0x33, 0x44])
        func make() -> (Data, AudioFormat) async throws -> Data {
            { [unowned self] data, format in
                self.calls.append((data, format))
                return self.pcmOut
            }
        }
    }

    // MARK: - Fixture loader

    /// Loads `transport-fixtures.json` from the repo-root `fixtures/` dir.
    /// For `CHAT_REQUEST_AUDIO_PCM16` the `data` field carries the literal
    /// `"<replaced-by-loader>"` placeholder; we substitute the base64 of the
    /// PCM16 fixture file before returning.
    private func loadFixture(_ key: String) throws -> [String: Any] {
        let url = fixturesURL().appendingPathComponent("transport-fixtures.json")
        let data = try Data(contentsOf: url)
        guard var root = try JSONSerialization.jsonObject(with: data) as? [String: Any],
              var fixture = root[key] as? [String: Any] else {
            XCTFail("fixture \(key) missing or not an object")
            return [:]
        }
        if key == "CHAT_REQUEST_AUDIO_PCM16" {
            let pcmURL = fixturesURL().appendingPathComponent("audio").appendingPathComponent("pcm16-1s-16khz-mono.bin")
            let pcmBytes = try Data(contentsOf: pcmURL)
            let b64 = pcmBytes.base64EncodedString()
            // Mutate messages[0].content[0].input_audio.data
            if var messages = fixture["messages"] as? [[String: Any]],
               var msg0 = messages.first,
               var parts = msg0["content"] as? [[String: Any]],
               var part0 = parts.first,
               var audio = part0["input_audio"] as? [String: Any] {
                audio["data"] = b64
                part0["input_audio"] = audio
                parts[0] = part0
                msg0["content"] = parts
                messages[0] = msg0
                fixture["messages"] = messages
                root[key] = fixture
            } else {
                XCTFail("CHAT_REQUEST_AUDIO_PCM16 fixture shape unexpected")
            }
        }
        return fixture
    }

    private func messages(from fixture: [String: Any]) -> [[String: Any]] {
        (fixture["messages"] as? [[String: Any]]) ?? []
    }

    private func fixturesURL() -> URL {
        var dir = URL(fileURLWithPath: #file).deletingLastPathComponent()
        while !FileManager.default.fileExists(atPath: dir.appendingPathComponent("fixtures").path) {
            let parent = dir.deletingLastPathComponent()
            if parent.path == dir.path {
                fatalError("fixtures dir not found walking up from \(#file)")
            }
            dir = parent
        }
        return dir.appendingPathComponent("fixtures")
    }

    // MARK: - Happy paths (driven by transport-fixtures.json)

    /// `CHAT_REQUEST_TEXT` — the legacy string-content shape produces a prompt
    /// with the user text and no media collateral.
    func testTextOnlyMessage() async throws {
        let fixture = try loadFixture("CHAT_REQUEST_TEXT")
        let translator = ContentPartsTranslator(mmprojLoaded: false, modelHasAudioEncoder: false)
        let result = try await translator.translate(messages: messages(from: fixture))
        XCTAssertEqual(result.prompt, "hi")
        XCTAssertTrue(result.media.isEmpty)
        XCTAssertEqual(result.messagesWithMarkers.count, 1)
        XCTAssertEqual(result.messagesWithMarkers[0].role, "user")
        XCTAssertEqual(result.messagesWithMarkers[0].content, "hi")
    }

    /// `CHAT_REQUEST_IMAGE` — text + data-URL image. The image part should be
    /// resolved via the (mocked) ImageDecoder and the bytes appended to
    /// `media`. The rendered content for that message has a single
    /// `<__media__>` marker substituted in place of the image part.
    func testTextPlusImage() async throws {
        let fixture = try loadFixture("CHAT_REQUEST_IMAGE")
        let mock = MockImageDecoder()
        let cannedPng = Data([0x89, 0x50, 0x4E, 0x47, 0x0D, 0x0A, 0x1A, 0x0A, 0x99])
        // Prefix-match any data: URL by snapping it after we observe it; here
        // we just set a default in `responses` keyed off the actual URL once
        // we know it from the fixture.
        let urlFromFixture: String = {
            let parts = (((fixture["messages"] as? [[String: Any]])?[0])?["content"] as? [[String: Any]]) ?? []
            return ((parts.first(where: { ($0["type"] as? String) == "image_url" })?["image_url"] as? [String: Any])?["url"] as? String) ?? ""
        }()
        mock.responses[urlFromFixture] = cannedPng

        let translator = ContentPartsTranslator(
            mmprojLoaded: true,
            modelHasAudioEncoder: false,
            imageDecoder: mock
        )
        let result = try await translator.translate(messages: messages(from: fixture))
        XCTAssertEqual(result.prompt, "What is in this image?")
        XCTAssertEqual(result.media.count, 1)
        XCTAssertEqual(result.media[0], cannedPng)
        XCTAssertEqual(mock.calls, [urlFromFixture])
        // Marker count in rendered content == media count.
        let markerCount = result.messagesWithMarkers
            .map { $0.content.components(separatedBy: MTMD_MEDIA_MARKER).count - 1 }
            .reduce(0, +)
        XCTAssertEqual(markerCount, 1)
    }

    /// `CHAT_REQUEST_AUDIO_PCM16` — base64-encoded PCM16 + text. The base64
    /// payload is fed (decoded) into the audio-decoder closure with
    /// `format == .pcm16`; the canned PCM result lands in `media`.
    func testAudioPCM16PlusText() async throws {
        let fixture = try loadFixture("CHAT_REQUEST_AUDIO_PCM16")
        let recorder = AudioRecorder()
        let translator = ContentPartsTranslator(
            mmprojLoaded: false,
            modelHasAudioEncoder: true,
            audioDecoder: recorder.make()
        )
        let result = try await translator.translate(messages: messages(from: fixture))
        XCTAssertEqual(result.prompt, "Transcribe this.")
        XCTAssertEqual(result.media.count, 1)
        XCTAssertEqual(result.media[0], recorder.pcmOut)
        XCTAssertEqual(recorder.calls.count, 1)
        XCTAssertEqual(recorder.calls[0].1, .pcm16)
        // The recorder receives the raw decoded PCM bytes (the fixture loader
        // base64-encodes the PCM file, the translator base64-decodes it back).
        let pcmFile = try Data(contentsOf: fixturesURL().appendingPathComponent("audio").appendingPathComponent("pcm16-1s-16khz-mono.bin"))
        XCTAssertEqual(recorder.calls[0].0, pcmFile)
        let markerCount = result.messagesWithMarkers
            .map { $0.content.components(separatedBy: MTMD_MEDIA_MARKER).count - 1 }
            .reduce(0, +)
        XCTAssertEqual(markerCount, 1)
    }

    /// Interleaved `[text, image, text, audio, text]` → media list preserves
    /// declaration order (image first, then audio); rendered content has
    /// exactly two `<__media__>` markers in the right positions.
    func testInterleavedTextImageAudio() async throws {
        let imageMock = MockImageDecoder()
        let imageBytes = Data([0xAA, 0xBB, 0xCC])
        imageMock.responses["data:image/png;base64,AAAA"] = imageBytes
        let audioRecorder = AudioRecorder()
        audioRecorder.pcmOut = Data([0x55, 0x66, 0x77])
        let translator = ContentPartsTranslator(
            mmprojLoaded: true,
            modelHasAudioEncoder: true,
            imageDecoder: imageMock,
            audioDecoder: audioRecorder.make()
        )
        let messages: [[String: Any]] = [[
            "role": "user",
            "content": [
                ["type": "text", "text": "before"],
                ["type": "image_url", "image_url": ["url": "data:image/png;base64,AAAA"]] as [String: Any],
                ["type": "text", "text": "between"],
                ["type": "input_audio", "input_audio": ["data": "AAAA", "format": "pcm16"]] as [String: Any],
                ["type": "text", "text": "after"],
            ],
        ]]
        let result = try await translator.translate(messages: messages)
        XCTAssertEqual(result.media.count, 2)
        XCTAssertEqual(result.media[0], imageBytes, "image must come first in declaration order")
        XCTAssertEqual(result.media[1], audioRecorder.pcmOut, "audio must come second")
        XCTAssertEqual(result.messagesWithMarkers.count, 1)
        let content = result.messagesWithMarkers[0].content
        let markerCount = content.components(separatedBy: MTMD_MEDIA_MARKER).count - 1
        XCTAssertEqual(markerCount, 2)
        // First marker should appear after "before" and before "between";
        // second after "between" and before "after".
        let firstMarker = content.range(of: MTMD_MEDIA_MARKER)!
        let secondMarker = content.range(of: MTMD_MEDIA_MARKER, range: firstMarker.upperBound..<content.endIndex)!
        let beforeRange = content.range(of: "before")!
        let betweenRange = content.range(of: "between")!
        let afterRange = content.range(of: "after")!
        XCTAssertLessThan(beforeRange.upperBound, firstMarker.lowerBound)
        XCTAssertLessThan(firstMarker.upperBound, betweenRange.lowerBound)
        XCTAssertLessThan(betweenRange.upperBound, secondMarker.lowerBound)
        XCTAssertLessThan(secondMarker.upperBound, afterRange.lowerBound)
    }

    // MARK: - Negative paths

    /// Image part with `mmprojLoaded == false` → `noMmprojForImage`. The
    /// translator must throw before even consulting the image decoder.
    func testImageWithoutMmprojThrows() async {
        let messages: [[String: Any]] = [[
            "role": "user",
            "content": [
                ["type": "image_url", "image_url": ["url": "data:image/png;base64,AAAA"]]
            ],
        ]]
        let mock = MockImageDecoder()
        let translator = ContentPartsTranslator(
            mmprojLoaded: false,
            modelHasAudioEncoder: false,
            imageDecoder: mock
        )
        do {
            _ = try await translator.translate(messages: messages)
            XCTFail("expected noMmprojForImage")
        } catch TranslatorError.noMmprojForImage {
            XCTAssertTrue(mock.calls.isEmpty, "translator should not invoke decoder when mmproj is missing")
        } catch {
            XCTFail("unexpected error: \(error)")
        }
    }

    /// Audio part with `modelHasAudioEncoder == false` → `audioWithoutAudioEncoder`.
    func testAudioWithoutEncoderThrows() async {
        let messages: [[String: Any]] = [[
            "role": "user",
            "content": [
                ["type": "input_audio", "input_audio": ["data": "AAAA", "format": "pcm16"]]
            ],
        ]]
        let translator = ContentPartsTranslator(mmprojLoaded: false, modelHasAudioEncoder: false)
        do {
            _ = try await translator.translate(messages: messages)
            XCTFail("expected audioWithoutAudioEncoder")
        } catch TranslatorError.audioWithoutAudioEncoder {
            // expected
        } catch {
            XCTFail("unexpected error: \(error)")
        }
    }

    /// Unsupported audio format (e.g. `vorbis`) → `unsupportedAudioFormat`
    /// with the offending format echoed back and the supported list filled in.
    func testUnsupportedAudioFormatThrows() async {
        let messages: [[String: Any]] = [[
            "role": "user",
            "content": [
                ["type": "input_audio", "input_audio": ["data": "AAAA", "format": "vorbis"]]
            ],
        ]]
        let translator = ContentPartsTranslator(mmprojLoaded: false, modelHasAudioEncoder: true)
        do {
            _ = try await translator.translate(messages: messages)
            XCTFail("expected unsupportedAudioFormat")
        } catch let TranslatorError.unsupportedAudioFormat(fmt, supported) {
            XCTAssertEqual(fmt, "vorbis")
            XCTAssertEqual(supported, ContentPartsTranslator.supportedAudioFormats)
            XCTAssertTrue(supported.contains("flac"), "iOS supported list should include flac")
        } catch {
            XCTFail("unexpected error: \(error)")
        }
    }

    /// Unknown content part type → `malformedRequest` with the offending type
    /// echoed in the message.
    func testUnknownContentPartTypeThrows() async {
        let messages: [[String: Any]] = [[
            "role": "user",
            "content": [
                ["type": "video_url", "video_url": ["url": "https://example.com/v.mp4"]]
            ],
        ]]
        let translator = ContentPartsTranslator(mmprojLoaded: true, modelHasAudioEncoder: true)
        do {
            _ = try await translator.translate(messages: messages)
            XCTFail("expected malformedRequest")
        } catch let TranslatorError.malformedRequest(reason) {
            XCTAssertTrue(reason.contains("video_url"), "expected reason to mention offending type, got: \(reason)")
        } catch {
            XCTFail("unexpected error: \(error)")
        }
    }

    /// Empty `input_audio.data` → `malformedRequest`. The audio decoder must
    /// not be invoked — this is a request-shape error caught before decode.
    func testEmptyAudioDataThrowsMalformedRequest() async {
        let translator = ContentPartsTranslator(
            mmprojLoaded: false,
            modelHasAudioEncoder: true,
            imageDecoder: MockImageDecoder(),
            audioDecoder: { _, _ in
                XCTFail("audio decoder should not be invoked for empty data")
                return Data()
            }
        )
        let messages: [[String: Any]] = [[
            "role": "user",
            "content": [[
                "type": "input_audio",
                "input_audio": ["data": "", "format": "pcm16"]
            ]]
        ]]
        do {
            _ = try await translator.translate(messages: messages)
            XCTFail("Expected throw")
        } catch TranslatorError.malformedRequest {
            // OK
        } catch {
            XCTFail("Unexpected: \(error)")
        }
    }

    /// Malformed base64 in `input_audio.data` → `malformedRequest` (not
    /// `audioDecodeFailed`). The audio decoder never runs — this is a
    /// pre-decode request-shape error.
    func testMalformedBase64ThrowsMalformedRequest() async {
        let translator = ContentPartsTranslator(
            mmprojLoaded: false,
            modelHasAudioEncoder: true,
            imageDecoder: MockImageDecoder(),
            audioDecoder: { _, _ in
                XCTFail("audio decoder should not be invoked for invalid base64")
                return Data()
            }
        )
        let messages: [[String: Any]] = [[
            "role": "user",
            "content": [[
                "type": "input_audio",
                "input_audio": ["data": "!!!not-valid-base64!!!", "format": "pcm16"]
            ]]
        ]]
        do {
            _ = try await translator.translate(messages: messages)
            XCTFail("Expected throw")
        } catch TranslatorError.malformedRequest {
            // OK
        } catch {
            XCTFail("Unexpected: \(error)")
        }
    }
}

File: packages/dvai-bridge-capacitor-llama/android/src/test/java/co/deepvoiceai/dvaibridge/llama/ContentPartsTranslatorTest.kt
package co.deepvoiceai.dvaibridge.llama

import kotlinx.coroutines.runBlocking
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.booleanOrNull
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.doubleOrNull
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.longOrNull
import org.junit.Assert.assertArrayEquals
import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Assert.fail
import org.junit.Test
import java.io.File
import java.util.Base64

class ContentPartsTranslatorTest {
    // region Mocks

    /** Image decoder that returns canned bytes per URL and records calls. */
    private class MockImageDecoder {
        val responses = mutableMapOf<String, ByteArray>()
        val calls = mutableListOf<String>()
        fun resolve(url: String): ByteArray {
            calls += url
            return responses[url] ?: byteArrayOf(0xDE.toByte(), 0xAD.toByte(), 0xBE.toByte(), 0xEF.toByte())
        }
    }

    /** Records every audio decode call's `(bytesIn, format)`. */
    private class AudioRecorder {
        val calls = mutableListOf<Pair<ByteArray, AudioFormat>>()
        var pcmOut: ByteArray = byteArrayOf(0x11, 0x22, 0x33, 0x44)
        fun decode(data: ByteArray, format: AudioFormat): ByteArray {
            calls += data to format
            return pcmOut
        }
    }

    // endregion

    // region Fixture loader

    /** Walks up from the gradle module dir until it finds the repo-root `fixtures/` dir. */
    private fun fixturesDir(): File {
        var dir = File(".").canonicalFile
        while (!File(dir, "fixtures").isDirectory) {
            dir = dir.parentFile ?: error("fixtures dir not found walking up from ${File(".").canonicalPath}")
        }
        return File(dir, "fixtures")
    }

    /**
     * Load a fixture out of `transport-fixtures.json` and return its `messages`
     * array as `List<Map<String, Any?>>` — the shape the translator consumes.
     * For `CHAT_REQUEST_AUDIO_PCM16` the placeholder `data` field is replaced
     * with the base64 of the `pcm16-1s-16khz-mono.bin` fixture before return.
     */
    private fun loadMessages(key: String): List<Map<String, Any?>> {
        val text = File(fixturesDir(), "transport-fixtures.json").readText(Charsets.UTF_8)
        val root = Json.parseToJsonElement(text).jsonObject
        val fixture = root[key]?.jsonObject ?: error("missing fixture: $key")
        val messagesArray = fixture["messages"] as? JsonArray
            ?: error("fixture $key has no messages array")
        @Suppress("UNCHECKED_CAST")
        val messages = messagesArray.map { (it.toAny() as Map<String, Any?>).toMutableMap() }
            .toMutableList()
        if (key == "CHAT_REQUEST_AUDIO_PCM16") {
            // Replace the placeholder `data` field on the audio part with
            // base64(PCM16 fixture file) so the translator's base64-decode
            // round-trips back to the raw PCM bytes.
            val pcmBytes = File(fixturesDir(), "audio/pcm16-1s-16khz-mono.bin").readBytes()
            val b64 = Base64.getEncoder().encodeToString(pcmBytes)
            val msg0 = (messages[0]).toMutableMap()
            @Suppress("UNCHECKED_CAST")
            val parts = (msg0["content"] as List<Map<String, Any?>>).map { it.toMutableMap() }.toMutableList()
            val part0 = parts[0]
            @Suppress("UNCHECKED_CAST")
            val audioObj = (part0["input_audio"] as Map<String, Any?>).toMutableMap()
            audioObj["data"] = b64
            part0["input_audio"] = audioObj
            parts[0] = part0
            msg0["content"] = parts
            messages[0] = msg0
        }
        return messages
    }

    /**
     * Convert a `JsonElement` tree to a vanilla Kotlin `Any?` tree
     * (`Map<String, Any?>`, `List<Any?>`, `String`, `Number`, `Boolean`, `null`)
     * — the shape the translator's `translate()` consumes.
     */
    private fun JsonElement.toAny(): Any? = when (this) {
        is JsonObject -> this.mapValues { it.value.toAny() }
        is JsonArray -> this.map { it.toAny() }
        is JsonPrimitive -> {
            if (this.isString) this.contentOrNull
            else this.booleanOrNull ?: this.longOrNull ?: this.doubleOrNull ?: this.contentOrNull
        }
    }

    // endregion

    // region Happy paths

    /** `CHAT_REQUEST_TEXT` — legacy string content → prompt only. */
    @Test
    fun `text-only message produces prompt and no media`() = runBlocking {
        val translator = ContentPartsTranslator(mmprojLoaded = false, modelHasAudioEncoder = false)
        val result = translator.translate(loadMessages("CHAT_REQUEST_TEXT"))
        assertEquals("hi", result.prompt)
        assertTrue(result.media.isEmpty())
        assertEquals(1, result.messagesWithMarkers.size)
        assertEquals("user", result.messagesWithMarkers[0].role)
        assertEquals("hi", result.messagesWithMarkers[0].content)
    }

    /** `CHAT_REQUEST_IMAGE` — text + data-URL image. */
    @Test
    fun `text plus image part populates media via decoder`() = runBlocking {
        val messages = loadMessages("CHAT_REQUEST_IMAGE")
        @Suppress("UNCHECKED_CAST")
        val parts = (messages[0]["content"] as List<Map<String, Any?>>)
        @Suppress("UNCHECKED_CAST")
        val urlFromFixture = (parts.first { it["type"] == "image_url" }["image_url"] as Map<String, Any?>)["url"] as String

        val mock = MockImageDecoder().apply {
            responses[urlFromFixture] = byteArrayOf(0x89.toByte(), 0x50, 0x4E, 0x47, 0x0D, 0x0A, 0x1A, 0x0A, 0x99.toByte())
        }
        val translator = ContentPartsTranslator(
            mmprojLoaded = true,
            modelHasAudioEncoder = false,
            imageDecoder = { url -> mock.resolve(url) },
        )
        val result = translator.translate(messages)
        assertEquals("What is in this image?", result.prompt)
        assertEquals(1, result.media.size)
        assertArrayEquals(mock.responses[urlFromFixture], result.media[0])
        assertEquals(listOf(urlFromFixture), mock.calls)
        val markerCount = result.messagesWithMarkers
            .sumOf { it.content.split(MTMD_MEDIA_MARKER).size - 1 }
        assertEquals(1, markerCount)
    }

    /**
     * `CHAT_REQUEST_AUDIO_PCM16` — base64 audio + text. The base64 payload is
     * decoded and the **raw bytes** land in `media` unchanged; mtmd does its
     * own format detection downstream via miniaudio, so the translator no
     * longer routes audio through `AudioDecoder`. The `audioDecoder`
     * collaborator is wired up but should not be invoked.
     */
    @Test
    fun `audio part is base64-decoded and passed raw to media without audio decoder`() = runBlocking {
        val messages = loadMessages("CHAT_REQUEST_AUDIO_PCM16")
        val recorder = AudioRecorder()
        val translator = ContentPartsTranslator(
            mmprojLoaded = false,
            modelHasAudioEncoder = true,
            audioDecoder = { data, fmt -> recorder.decode(data, fmt) },
        )
        val result = translator.translate(messages)
        assertEquals("Transcribe this.", result.prompt)
        assertEquals(1, result.media.size)
        // `media[0]` is the raw base64-decoded fixture bytes — NOT the canned
        // `recorder.pcmOut` — because the translator no longer routes audio
        // through the decoder closure on the production path.
        val pcmFile = File(fixturesDir(), "audio/pcm16-1s-16khz-mono.bin").readBytes()
        assertArrayEquals(pcmFile, result.media[0])
        assertEquals("audioDecoder must not be called on the production path; mtmd handles decode itself", 0, recorder.calls.size)
        val markerCount = result.messagesWithMarkers
            .sumOf { it.content.split(MTMD_MEDIA_MARKER).size - 1 }
        assertEquals(1, markerCount)
    }

    /**
     * Interleaved [text, image, text, audio, text] → media list preserves
     * declaration order; rendered content has exactly two `<__media__>`
     * markers in the right positions. After the audio-path fix, `media[1]`
     * is the raw base64-decoded audio bytes; `audioDecoder` must not be
     * invoked.
     */
    @Test
    fun `interleaved text image audio preserves order`() = runBlocking {
        val imageMock = MockImageDecoder().apply {
            responses["data:image/png;base64,AAAA"] = byteArrayOf(0xAA.toByte(), 0xBB.toByte(), 0xCC.toByte())
        }
        val recorder = AudioRecorder().apply { pcmOut = byteArrayOf(0x55, 0x66, 0x77) }
        val translator = ContentPartsTranslator(
            mmprojLoaded = true,
            modelHasAudioEncoder = true,
            imageDecoder = { url -> imageMock.resolve(url) },
            audioDecoder = { data, fmt -> recorder.decode(data, fmt) },
        )
        val messages = listOf(
            mapOf<String, Any?>(
                "role" to "user",
                "content" to listOf(
                    mapOf("type" to "text", "text" to "before"),
                    mapOf("type" to "image_url", "image_url" to mapOf("url" to "data:image/png;base64,AAAA")),
                    mapOf("type" to "text", "text" to "between"),
                    mapOf("type" to "input_audio", "input_audio" to mapOf("data" to "AAAA", "format" to "pcm16")),
                    mapOf("type" to "text", "text" to "after"),
                ),
            ),
        )
        val result = translator.translate(messages)
        assertEquals(2, result.media.size)
        assertArrayEquals(byteArrayOf(0xAA.toByte(), 0xBB.toByte(), 0xCC.toByte()), result.media[0])
        // `"AAAA"` base64-decoded is three zero bytes — that's what mtmd sees.
        assertArrayEquals(byteArrayOf(0x00, 0x00, 0x00), result.media[1])
        assertEquals("audioDecoder must not be invoked on the production path", 0, recorder.calls.size)
        assertEquals(1, result.messagesWithMarkers.size)
        val content = result.messagesWithMarkers[0].content
        val markerCount = content.split(MTMD_MEDIA_MARKER).size - 1
        assertEquals(2, markerCount)
        // Position checks: first marker between "before" and "between";
        // second between "between" and "after".
        val firstMarker = content.indexOf(MTMD_MEDIA_MARKER)
        val secondMarker = content.indexOf(MTMD_MEDIA_MARKER, firstMarker + 1)
        val before = content.indexOf("before")
        val between = content.indexOf("between")
        val after = content.indexOf("after")
        assertTrue(before < firstMarker)
        assertTrue(firstMarker < between)
        assertTrue(between < secondMarker)
        assertTrue(secondMarker < after)
    }

    // endregion

    // region Negative paths

    /** Image part with `mmprojLoaded == false` → `NoMmprojForImage`. */
    @Test
    fun `image without mmproj throws and skips decoder`() = runBlocking {
        val mock = MockImageDecoder()
        val translator = ContentPartsTranslator(
            mmprojLoaded = false,
            modelHasAudioEncoder = false,
            imageDecoder = { url -> mock.resolve(url) },
        )
        val messages = listOf(
            mapOf<String, Any?>(
                "role" to "user",
                "content" to listOf(
                    mapOf("type" to "image_url", "image_url" to mapOf("url" to "data:image/png;base64,AAAA")),
                ),
            ),
        )
        try {
            translator.translate(messages)
            fail("expected NoMmprojForImage")
        } catch (e: TranslatorError.NoMmprojForImage) {
            assertTrue("decoder should not be called when mmproj is missing", mock.calls.isEmpty())
        }
        Unit
    }

    /** Audio part with `modelHasAudioEncoder == false` → `AudioWithoutAudioEncoder`. */
    @Test
    fun `audio without encoder throws`() = runBlocking {
        val translator = ContentPartsTranslator(mmprojLoaded = false, modelHasAudioEncoder = false)
        val messages = listOf(
            mapOf<String, Any?>(
                "role" to "user",
                "content" to listOf(
                    mapOf("type" to "input_audio", "input_audio" to mapOf("data" to "AAAA", "format" to "pcm16")),
                ),
            ),
        )
        try {
            translator.translate(messages)
            fail("expected AudioWithoutAudioEncoder")
        } catch (_: TranslatorError.AudioWithoutAudioEncoder) {
            // expected
        }
        Unit
    }

    /** Unsupported audio format (e.g. `vorbis`) → `UnsupportedAudioFormat`. */
    @Test
    fun `unsupported audio format throws with platform list`() = runBlocking {
        val translator = ContentPartsTranslator(mmprojLoaded = false, modelHasAudioEncoder = true)
        val messages = listOf(
            mapOf<String, Any?>(
                "role" to "user",
                "content" to listOf(
                    mapOf("type" to "input_audio", "input_audio" to mapOf("data" to "AAAA", "format" to "vorbis")),
                ),
            ),
        )
        try {
            translator.translate(messages)
            fail("expected UnsupportedAudioFormat")
        } catch (e: TranslatorError.UnsupportedAudioFormat) {
            assertEquals("vorbis", e.format)
            assertEquals(ContentPartsTranslator.SUPPORTED_AUDIO_FORMATS, e.supported)
            assertTrue("Android supported list should include ogg", e.supported.contains("ogg"))
        }
        Unit
    }

    /** Unknown content part type → `MalformedRequest` mentioning the type. */
    @Test
    fun `unknown content part type throws MalformedRequest`() = runBlocking {
        val translator = ContentPartsTranslator(mmprojLoaded = true, modelHasAudioEncoder = true)
        val messages = listOf(
            mapOf<String, Any?>(
                "role" to "user",
                "content" to listOf(
                    mapOf("type" to "video_url", "video_url" to mapOf("url" to "https://example.com/v.mp4")),
                ),
            ),
        )
        try {
            translator.translate(messages)
            fail("expected MalformedRequest")
        } catch (e: TranslatorError.MalformedRequest) {
            assertTrue("expected message to mention 'video_url', got: ${e.message}", (e.message ?: "").contains("video_url"))
        }
        Unit
    }

    /**
     * Empty `input_audio.data` → `MalformedRequest`. The audio decoder must
     * not be invoked — this is a request-shape error caught before decode.
     */
    @Test
    fun `empty audio data throws MalformedRequest`() = runBlocking {
        val translator = ContentPartsTranslator(
            mmprojLoaded = false,
            modelHasAudioEncoder = true,
            imageDecoder = { _ -> error("image decoder should not be called") },
            audioDecoder = { _, _ -> error("audio decoder should not be called for empty data") },
        )
        val messages = listOf(
            mapOf<String, Any?>(
                "role" to "user",
                "content" to listOf(
                    mapOf("type" to "input_audio", "input_audio" to mapOf("data" to "", "format" to "pcm16")),
                ),
            ),
        )
        try {
            translator.translate(messages)
            fail("expected MalformedRequest")
        } catch (_: TranslatorError.MalformedRequest) {
            // expected
        }
        Unit
    }

    /**
     * Malformed base64 in `input_audio.data` → `MalformedRequest` (not
     * `AudioDecodeFailed`). The audio decoder never runs — this is a
     * pre-decode request-shape error.
     */
    @Test
    fun `malformed base64 audio data throws MalformedRequest`() = runBlocking {
        val translator = ContentPartsTranslator(
            mmprojLoaded = false,
            modelHasAudioEncoder = true,
            imageDecoder = { _ -> error("image decoder should not be called") },
            audioDecoder = { _, _ -> error("audio decoder should not be called for invalid base64") },
        )
        val messages = listOf(
            mapOf<String, Any?>(
                "role" to "user",
                "content" to listOf(
                    mapOf(
                        "type" to "input_audio",
                        "input_audio" to mapOf("data" to "!!!not-valid-base64!!!", "format" to "pcm16"),
                    ),
                ),
            ),
        )
        try {
            translator.translate(messages)
            fail("expected MalformedRequest")
        } catch (_: TranslatorError.MalformedRequest) {
            // expected
        }
        Unit
    }

    // endregion
}
